src/common/reactNativeProjectHelper.ts

```diff
@@ -10,53 +10,111 @@ import {Package} from "./node/package";
             });
     }
 
-    public static getReactNativePackageVersionFromNodeModules(projectRoot: string): Q.Promise<string> {
-        return new Package(projectRoot).dependencyPackage("react-native").version()
-            .then(version => ReactNativeProjectHelper.processVersion(version))
-            .catch(err => {
-                throw ErrorHelper.getInternalError(InternalErrorCode.ReactNativePackageIsNotInstalled);
-            });
+    public static getReactNativePackageVersionsFromNodeModules(projectRoot: string, isRNWVersion: boolean = false): Q.Promise<RNPackageVersions> {
+        let versionPromises: Q.Promise<string>[] = [];
+
+        versionPromises.push(
+            new Package(projectRoot).getPackageVersionFromNodeModules("react-native")
+                .then(version => ReactNativeProjectHelper.processVersion(version, true))
+                .catch(err => {
+                    throw ErrorHelper.getInternalError(InternalErrorCode.ReactNativePackageIsNotInstalled);
+                })
+            );
+
+        if (isRNWVersion) {
+            versionPromises.push(
+                new Package(projectRoot).getPackageVersionFromNodeModules("react-native-windows")
+                    .then(version => ReactNativeProjectHelper.processVersion(version, false))
+                    .catch(err => "")
+            );
+        }
+
+        return Q.all(versionPromises).then(packageVersions => ({
+            reactNativeVersion: packageVersions[0],
+            reactNativeWindowsVersion: packageVersions[1] || "",
+        }));
     }
 
-    public static getReactNativeVersionFromProjectPackage(cwd: string): Q.Promise<string> {
-        const rootProjectPackageJson = new Package(cwd);
-        return rootProjectPackageJson.dependencies()
-            .then(dependencies => {
-                if (dependencies["react-native"]) {
-                    return ReactNativeProjectHelper.processVersion(dependencies["react-native"]);
-                }
-                return rootProjectPackageJson.devDependencies()
-                    .then(devDependencies => {
-                        if (devDependencies["react-native"]) {
-                            return ReactNativeProjectHelper.processVersion(devDependencies["react-native"]);
+    public static getReactNativeVersionsFromProjectPackage(cwd: string, isRNWVersion: boolean = false): Q.Promise<RNPackageVersions> {
+        let parsedPackageNames: ParsedPackageName[] = [
+            {
+                packageName: "react-native",
+                useSemverCoerce: true,
+            },
+        ];
+
+        if (isRNWVersion) {
+            parsedPackageNames.push({
+                packageName: "react-native-windows",
+                useSemverCoerce: false,
+            });
+        }
+
+        let versionPromises: Q.Promise<string>[] = [];
+
+        parsedPackageNames.forEach(parsedPackageName => {
```

Comment:
As we parse only one `package.json` file, can we parse it once and get all needed dependencies from there? For example we can try to cache dependencies in variable and then get what we need.

Reply:
Yes, we can cache dependencies and then get required versions. Such approach will be more efficient.